x-pack/plugins/cases/public/components/observables/observable_form.tsx

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { type FC, useCallback, useMemo, memo, useState } from 'react';
+import {
+  useForm,
+  Form,
+  UseField,
+  useFormContext,
+} from '@kbn/es-ui-shared-plugin/static/forms/hook_form_lib';
+import { EuiButton, EuiFlexGroup, EuiSpacer } from '@elastic/eui';
+
+import {
+  TextAreaField,
+  SelectField,
+  TextField,
+} from '@kbn/es-ui-shared-plugin/static/forms/components';
+
+import { OBSERVABLE_TYPES_BUILTIN } from '../../../common/constants';
+import type { ObservablePatchType, ObservablePostType } from '../../../common/types/api';
+import type { Observable } from '../../../common/types/domain';
+import { useGetCaseConfiguration } from '../../containers/configure/use_get_case_configuration';
+import * as i18n from './translations';
+import { fieldsConfig, normalizeValueType } from './fields_config';
+
+export interface ObservableFormFieldsProps {
+  observable?: Observable;
+}
+
+export const ObservableFormFields = memo(({ observable }: ObservableFormFieldsProps) => {
+  const { data, isLoading } = useGetCaseConfiguration();
+  const [selectedTypeKey, setSelectedTypeKey] = useState<string>(observable?.typeKey ?? '');
+
+  const { validateFields } = useFormContext();
+
+  const options = useMemo(() => {
+    return [...OBSERVABLE_TYPES_BUILTIN, ...data.observableTypes].map((observableType) => ({
+      value: observableType.key,
+      text: observableType.label,
+    }));
+  }, [data.observableTypes]);
+
+  const handleSelectedTypeChange = useCallback(
+    (selectedTypeKeyValue: string) => {
+      validateFields(['value']);
+      setSelectedTypeKey(selectedTypeKeyValue);
+    },
+    [validateFields]
+  );
+
+  return (
+    <>
+      {!observable && (
+        <UseField
+          component={SelectField}
+          componentProps={{ euiFieldProps: { options, hasNoInitialSelection: true, isLoading } }}
+          onChange={handleSelectedTypeChange}
+          path="typeKey"
+          config={fieldsConfig.typeKey}
+        />
+      )}
+      <UseField
+        path="value"
+        config={fieldsConfig.value[normalizeValueType(selectedTypeKey)]}
```

Comment:
After its first render, you cannot change the `config` prop of the `UseField` component. It will have the initial config. I would suggest instead of having a factory for the configs to have a factory for components. Each observable type will have its own component that renders a `UseField` with its own config inside the component. Then you can use the component factory to render the proper component based on the selection of the user. You can take a look of how we do it here `x-pack/plugins/cases/public/components/custom_fields/form_fields.tsx`.

Reply:
just calling validation on field change again